mixer/pkg/il/evaluator/expressionbuilder.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package evaluator
+
+import (
+	"istio.io/istio/mixer/pkg/attribute"
+	"istio.io/istio/mixer/pkg/expr"
+	"istio.io/istio/mixer/pkg/il/compiler"
+	"istio.io/istio/mixer/pkg/il/interpreter"
+	"istio.io/istio/mixer/pkg/il/runtime"
+)
+
+// ExpressionBuilder is used to create a set of pre-compiled expressions, backed by the same program and interpreter
+// instance. It is meant to be used to create a large number of precompiled expressions that are backed by an efficient
+// set of shared, immutable objects.
+type ExpressionBuilder struct {
+	compiler    *compiler.Compiler
+	interpreter *interpreter.Interpreter
+}
+
+// NewBuilder returns a new ExpressionBuilder
+func NewBuilder(finder expr.AttributeDescriptorFinder) *ExpressionBuilder {
+	return newBuilder(finder, allFunctions, allExterns)
+}
+
+func newBuilder(finder expr.AttributeDescriptorFinder, functions map[string]expr.FunctionMetadata, externs map[string]interpreter.Extern) *ExpressionBuilder {
+	c := compiler.New(finder, functions)
+	return &ExpressionBuilder{
+		compiler:    c,
+		interpreter: interpreter.New(c.Program(), externs),
+	}
+}
+
+// Compile the given text and return a pre-compiled expression object.
+func (e *ExpressionBuilder) Compile(text string) (Expression, error) {
+	fnID, err := e.compiler.CompileExpression(text)
+	if err != nil {
+		return nil, err
+	}
+
+	return &expression{
```

Comment:
You told me about the perf difference between using Expression as an interface vs. changing the return type to effectively &expression directly, and that for some magical reason returning an interface is faster.

But have you tried returning an expression struct as-is? Not the address of the struct, the struct itself, and changing all places that hold a pointer to the struct to instead hold the struct proper? I bet that would be faster and generate less garbage.

Reply:
I tried various combinations again, and this time around the numbers seemed pretty much similar. (There is ~10ns/op variance in the numbers, easy)

// 12/6/2017 use direct struct 
(i.e. type Expression struct{}, foo := Expression{})
//BenchmarkCompiled/ExprBench/ok_1st-8          	10000000	       167 ns/op	       0 B/op	       0 allocs/op
//BenchmarkCompiled/ExprBench/ok_2nd-8          	 5000000	       272 ns/op	      16 B/op	       1 allocs/op
//BenchmarkCompiled/ExprBench/not_found-8       	 5000000	       278 ns/op	      16 B/op	       1 allocs/op

// 12/6/2017 use struct pointer
 (i.e. type Expression struct{}, foo := &Expression{})
//BenchmarkCompiled/ExprBench/ok_1st-8          	10000000	       165 ns/op	       0 B/op	       0 allocs/op
//BenchmarkCompiled/ExprBench/ok_2nd-8          	 5000000	       268 ns/op	      16 B/op	       1 allocs/op
//BenchmarkCompiled/ExprBench/not_found-8       	 5000000	       272 ns/op	      16 B/op	       1 allocs/op

// 12/6/2017 use direct struct with interface 
(i.e. type Expression interface, var _ Expression := expression{})
//BenchmarkCompiled/ExprBench/ok_1st-8          	10000000	       154 ns/op	       0 B/op	       0 allocs/op
//BenchmarkCompiled/ExprBench/ok_2nd-8          	 5000000	       259 ns/op	      16 B/op	       1 allocs/op
//BenchmarkCompiled/ExprBench/not_found-8       	 5000000	       258 ns/op	      16 B/op	       1 allocs/op

// Best case for the last experiment:
//BenchmarkCompiled/ExprBench/ok_1st-8          	10000000	       149 ns/op	       0 B/op	       0 allocs/op
//BenchmarkCompiled/ExprBench/ok_2nd-8          	 5000000	       254 ns/op	      16 B/op	       1 allocs/op
//BenchmarkCompiled/ExprBench/not_found-8       	 5000000	       254 ns/op	      16 B/op	       1 allocs/op

So, using naked, no-pointer structs seems to be worth it. Using no-pointer structs behind an interface seems to be even better. :)

I'll transition the code to use naked structs, no-pointers. This should help the garbage collector. Thanks for pointing out.